### DISCUSSION TITLE
6/8 pattern question - R1.1.1-Dec 6, 2021 version

### DISCUSSION THREAD

**Discussion by gvergis** [hydrogen-music/hydrogen#1433] [Hydrogen on Linux]
I am trying to record a song in Ardour with a 6/8 beat pattern in Hydrogen version 1.1.1-Dec 6th 2021. 

In the hydrogen pattern editor, I have set the size field to 6/8 and the Res field to 1/8. However, I am only seeing 3 beats in the pattern editor, while I was expecting to see six beats. To see all six beats in the pattern editor, I have to double the numerator to 12 (Size field has to be set to 12/8). For 4/4 and 3/4 patterns, I get the right number  of beats, without the need to manipulate the size. So it is only the 6/8 that is forcing me to double the value. Going through some older discussion on this topic, I thought this issue is already fixed in the new version -- the number of beats you get is what you set in the size field, no more, no less.

Meanwhile, I have set Ardour to 6/8 beat pattern (there the metronome clicks correctly, etc). Now, I have made Ardour the Jack transport time master. In Hydrogen, Artificially doubling the size to get the right number of beats in a measure leads to wrong transport position in Hydrogen, especially when Ardour is set to 6/8. In this example, Ardour would have advanced a whole measure while Hydrogen is only half way through the measure, as the Jack transport attempts to synchronize the two programs. 

Am I missing something with my understanding of the pattern editor in Hydrogen? I haven't investigated other beat pattern values, besides the common 4/4, 3/4, and 6/8. Both 4/4 and 3/4 work intuitively and correctly. Please advice.

**Comment by cme** [comment]
The new pattern length setting UI is probably a bit misleading, in that while it looks like it sets a time signature, it only really sets the pattern *length* in quarter-notes. Perhaps this should change and an actual "time signature" concept be introduced.

The pattern ruler numbers always number quarter-notes (so for a pattern that's set as "8/8" in the "length" UI, it will still show "1 2 3 4" where it should probably show "1 2 3 4 5 6 7 8"), ie. the beats of the *tempo* (which is always considered as quarter-notes, at least in MIDI) rather than the beats of the time signature. Numbering the beats of the time signature would probably be more useful, or perhaps numbering both (or a visible option to select which) is probably something that would be useful.

For the moment, for a true 6/8 you'll have to (somewhat painfully, I'm afraid) ignore the numbers and work just in 1/8th notes. 

<img width="531" alt="image" src="https://user-images.githubusercontent.com/768466/147822690-ab92a588-c179-40d4-a873-f23404752c75.png">

@oddtime , @theGreatWhiteShark , have I misrepresented anything? Should we consider changing the ruler (and pattern markings) in response to the length denominator, or any other thoughts?

**Reply by oddtime** [reply]
The same affects also the painted grid lines.
For 6/8 the standard grid should be have the thicker line in the middle of the bar.
For times in 16th notes, painting each 16th note is not a good solution since the ruler would be full of overlapping text.
In case of a 17/16 it could be thought as (4+5+4+4)/16 and painted '1', '5', '10', '14', but also (5+4+4+4)/16 or any subdivision.
User may need multiple subdivisions for 8th notes meters as well (e.g. 9/8 as 3+3+3 or 5+4 or 4+5 or 2+2+2+2+1 in a balcan style)

I think that custom beams in the grid would solve the problem.
I don't know if there is a smarter way of automate the choice of "what numbers to print in the ruler".

**Reply by oddtime** [reply]
But on the other hand grid lines depend on the selected resolution dynamically.
If one uses tuplets (I am thinking to my experimental branch) the thickness of painted lines must respect the tuplet and not the meter, and the beam lengths according to the tuplet and the meter may differ of course

**Reply by gvergis** [reply]
> The new pattern length setting UI is probably a bit misleading, in that while it looks like it sets a time signature, it only really sets the pattern _length_ in quarter-notes. Perhaps this should change and an actual "time signature" concept be introduced.
> 
> The pattern ruler numbers always number quarter-notes (so for a pattern that's set as "8/8" in the "length" UI, it will still show "1 2 3 4" where it should probably show "1 2 3 4 5 6 7 8"), ie. the beats of the _tempo_ (which is always considered as quarter-notes, at least in MIDI) rather than the beats of the time signature. Numbering the beats of the time signature would probably be more useful, or perhaps numbering both (or a visible option to select which) is probably something that would be useful.
> 
> For the moment, for a true 6/8 you'll have to (somewhat painfully, I'm afraid) ignore the numbers and work just in 1/8th notes.

Do you mean setting the size to 6/8 will set the actual bar length to 6 beats, as seen by the external Jack time master? If so, I agree I will just work on 1/8 note grid.

I like the idea of "pattern entry mode" where actual musical meter (time signature) is an option along with the existing size parameter. If the mode can be configured in the user preference or toggled in the pattern editor (mouse click toggles between the two options dynamically), it would satisfy all users?

Ardour, all DAW for that matter, represents content in  the BBT ruler [ Bar | Beat | Time ]. Eventually, having the [song.pattern.beat] position in hydrogen visually match to the external jack time master consistently helps song editing process tremendously. 

As a future feature request, it would be useful if the hydrogen display continually tracks the actual pattern in the song. I believe it currently only shows whatever pattern that was selected and not automatically update to the pattern that the song position calls for.



> The same affects also the painted grid lines. For 6/8 the standard grid should be have the thicker line in the middle of the bar. For times in 16th notes, painting each 16th note is not a good solution since the ruler would be full of overlapping text. In case of a 17/16 it could be thought as (4+5+4+4)/16 and painted '1', '5', '10', '14', but also (5+4+4+4)/16 or any subdivision. User may need multiple subdivisions for 8th notes meters as well (e.g. 9/8 as 3+3+3 or 5+4 or 4+5 or 2+2+2+2+1 in a balcan style)
> 
> I think that custom beams in the grid would solve the problem. I don't know if there is a smarter way of automate the choice of "what numbers to print in the ruler".

Personally, the grids lines (tic narks without any emphasis) based on selected Resolution looks great, and intuitive to me -:)

**Reply by theGreatWhiteShark** [reply]
> The new pattern length setting UI is probably a bit misleading, in that while it looks like it sets a time signature, it only really sets the pattern length in quarter-notes. Perhaps this should change and an actual "time signature" concept be introduced.

Definitely! Since I don't know much about music theory, I have no idea what's the difference between 6/8 and 3/4. But we should use time signature because this is most probably what people would expect a music software to use. 

> Should we consider changing the ruler (and pattern markings) in response to the length denominator, or any other thoughts?

Hmm. Sounds like the right thing to do.

**Reply by oddtime** [reply]
Sorry I meant "darker" lines, not "thicker". (as it is now indeed)

**Reply by cme** [reply]
> As a future feature request, it would be useful if the hydrogen display continually tracks the actual pattern in the song. I believe it currently only shows whatever pattern that was selected and not automatically update to the pattern that the song position calls for.

Do you mean having the Song Editor follow the playing patterns (by scrolling down), or having the Pattern Editor change to follow the current 'current' pattern? Both have been suggested in the past but have the problem that there's no single 'current' pattern in a song because at any point in a song, any number of patterns may be actively playing at the same time. I'm working on a solution for the Song Editor which should cover most use-cases. Following the 'current' pattern in the Song Editor has a couple of potential behaviours that would satisfy what people are looking for in this, so I will experiment with those too.

**Reply by gvergis** [reply]
Yes, I meant the song editor following the current pattern. You are right, I overlooked the case where there are multiple patterns at a particular song position. As an example of the case where I have two patterns in the same measure of a song  is where I insert a pattern for the fill, followed by another pattern with a crash cymbal in the first beat of  the next measure, where there would be a regular beat pattern, as well. Clearly an interesting dilemma on which pattern to prioritize for displaying! Having a composite view of all active notes from all active patterns is an overkill and not very justifiable effort (nice to have, not a must have).

**Reply by cme** [reply]
> Having a composite view of all active notes from all active patterns is an overkill and not very justifiable effort (nice to have, not a must have).

Well, funny you mention that... #1310 ...

**Comment by theGreatWhiteShark** [comment]
> Meanwhile, I have set Ardour to 6/8 beat pattern (there the metronome clicks correctly, etc). Now, I have made Ardour the Jack transport time master. In Hydrogen, Artificially doubling the size to get the right number of beats in a measure leads to wrong transport position in Hydrogen, especially when Ardour is set to 6/8. In this example, Ardour would have advanced a whole measure while Hydrogen is only half way through the measure, as the Jack transport attempts to synchronize the two programs.

I have to admit, JACK timebase support is quite a pain right now. Basically, transport control in JACK and Hydrogen are not fully compatible right now  and TBM sync is only supported some workarounds. It should work when you either alter only time signatures or tempo in the timebase master and set the corresponding option in the Hydrogen preferences. Or best, just turn off timebase support in the Preferences. I'm working on a fix for this right now, which will be part of the next release.

> Ardour, all DAW for that matter, represents content in the BBT ruler [ Bar | Beat | Time ]. Eventually, having the [song.pattern.beat] position in hydrogen visually match to the external jack time master consistently helps song editing process tremendously.

Well, this is one of the two problems the JACK TBM support is such a mess. What's a "Bar"? In applications like Ardour you have to set a time signature and the position of all your bars in the continuous transport are a result of your tempo, tempo changes, and time signature changes throughout the song. Hydrogen, on the other hand, was designed to work quite differently. Here, you create patterns of specific lengths and arrange them in the song editor. There can be even holes with no patterns present at all (using a tempo-dependent default length of silence instead), you are able to toggle patterns during playback (and alter the length of a part already passed), and there is no indefinite transport (unless you loop transport). With TBM active the JACK server sends a specific time signature to use at the current transport position. Using the mapping of pattern length to time signature as discussed above, this would allow the JACK timebase master to alter pattern lengths in Hydrogen. This doesn't feel right at all and I think we'll deprecate TBM support before allowing an external application to change patterns. 

In general a  [ Bar | Beat | Tick ] display of the transport position would be nice. But if we don't get an intuitive meaning of Bars that match the one used in JACK, this would only cause confusion. Any ideas this could work?

**Reply by gvergis** [reply]
Introducing BBT as a new mode when Jack TBM is used would be a solution that preserves existing code and MIDI semantics. A few suggestions for the new mode:

1) Rename the "Size" label to "Beat" or "Meter" or "Rhythm", to indicate that it is a true musical meter. The numerator value is used to enumerate the beat marker. In case of 6/8, there will be 6 numbered beat markers.
2) Rename the "Res." label to "Note". Denominator  of the musical meter sets the minimum resolution (note value). That means, 1/4 will be removed (grayed out) as an option when 6/8 meter is chosen. Effectively, denominator of resolution >= denominator of the meter. I am not entirely sure if this is a good idea or not (so open to debate on this from a music theory or a practical usage perspective!)

During the music arrangement phase, the end user is expected to ensure that the measures (bars) have the same time signature and tempo between Ardour and Hydrogen, for each and every measure in the song. This "composition/arrangement' requirement applies even to rest/silence measures. Such "arrangement" ensures that the transport of each program is synchronized [Transport_Time_position=Note_value * Song_Measure_number/BPM]. Once this requirement is fulfilled, Ardour TBM "should" be able to exactly locate the correct "note" in Hydrogen. 

A few note on my usage model: For a new session, I typically start by putting a beat down in Hydrogen. I then build a Cue track in Ardour while listening to the beats rendered in Hydrogen with Ardour as a Jack TBM. During the Cue track building process, I often add a measure of rest or a drum fill in Hydrogen, say between a verse/chorus transition. This maily because I am trying to compose an interlude instrumental music. Hence the need to have the transport show the same position in both programs. I also extensively use the Hydrogen "Director" to ensure that the count is "Note timing" accurate, and as a cue to punch in / punch out a take in various Audio programs (Rosegarden, Ardour, Muse, etc).

**Reply by oddtime** [reply]
>  That means, 1/4 will be removed (grayed out) as an option when 6/8 meter is chosen. Effectively, denominator of resolution >= denominator of the meter. I am not entirely sure if this is a good idea or not (so open to debate on this from a music theory or a practical usage perspective!)

In my opinion all items should be active (more powerful) and I don't see a theoretical reason to limit the choice (6/8 and 3/4 coexist in afro).

Regarding the names of labels, the current ones seem more exact to me actually... "Meter" or "Rhythm" do not mean how much long is the pattern (think of a pattern of 4 bars in 4/4, meter is 4/4 while length 16/4), while "Beat" is unclear to me.
"Note" alone does not give a sense of "quantum" length, while "Res." does.

**Reply by oddtime** [reply]
Another thought: hydrogen fully supports times in 12th notes (I don't think that this happens in ardour) plus custom fractional numerator (not fully yet)...

**Reply by theGreatWhiteShark** [reply]
> Introducing BBT as a new mode when Jack TBM is used

Do you mean UI/UX should change once the JACK server starts/stops providing BBT information?

It would be best if things stay consistent. Else the UX would become less intuitive and the code base harder to maintain. In addition, most of Hydrogen users (according to a poll some years ago) use Windows. Of those using Linux not all are using the JACK server and of those probably only a fraction make use of the additional BBT support. We should be careful changing the UI/UX for everyone just to make a small feature more easy to use.

> During the music arrangement phase, the end user is expected to ensure that the measures (bars) have the same time signature and tempo between Ardour and Hydrogen, for each and every measure in the song. This "composition/arrangement' requirement applies even to rest/silence measures. Such "arrangement" ensures that the transport of each program is synchronized [Transport_Time_position=Note_value * Song_Measure_number/BPM]. Once this requirement is fulfilled, Ardour TBM "should" be able to exactly locate the correct "note" in Hydrogen.

This is actually the situation we are facing right now. But when the user is required to keep time signatures and tempo markers consistent between Hydrogen and the timebase master, just regular JACK transport without BBT is sufficient to maintain sync (at least as far as I see). On the other hand, when requiring this level of manual work - without Hydrogen being able to check and tell whether things are consistent - it can hardly be called a feature. Especially since Hydrogen is not helpful when comparing time signatures (all squares in the song editor have the same size). I would love to have a more easy/automated way of keeping those things in sync. But i can't see how right now.

**Reply by gvergis** [reply]
>It would be best if things stay consistent. Else the UX would become less intuitive and the code base harder to maintain. In addition, most of Hydrogen users (according to a poll some years ago) use Windows. Of those using Linux not all are using the JACK server and of those probably only a fraction make use of the additional BBT support. We should be careful changing the UI/UX for everyone just to make a small feature more easy to use.

Completely agree, I overlooked the windows environment.

>when the user is required to keep time signatures and tempo markers consistent between Hydrogen and the timebase master, just regular JACK transport without BBT is sufficient to maintain sync (at least as far as I see).

Agreed.

>manual work - without Hydrogen being able to check and tell whether things are consistent - it can hardly be called a feature. Especially since Hydrogen is not helpful when comparing time signatures (all squares in the song editor have the same size). 

So true. I often wished there was a better way. Right now, the only indication of inconsistent time signatures is audible out-of-sync beats

**Comment by gvergis** [comment]
>The same affects also the painted grid lines. For 6/8 the standard grid should be have the thicker line in the middle of the bar. For times in 16th notes, painting each 16th note is not a good solution since the ruler would be full of overlapping text. 

>In case of a 17/16 it could be thought as (4+5+4+4)/16 and painted '1', '5', '10', '14', but also (5+4+4+4)/16 or any subdivision. User may need multiple subdivisions for 8th notes meters as well (e.g. 9/8 as 3+3+3 or 5+4 or 4+5 or 2+2+2+2+1 in a balcan style)

Would a custom sub-division field help solve this dilemma? The proposal would be to have a field that the grid painter uses. The field would have some comma separated ordered numbers. In the 17/16 example above, one would enter (4,5,4,4) or any other sub-division of choice.

>I think that custom beams in the grid would solve the problem. I don't know if there is a smarter way of automate the choice of "what numbers to print in the ruler".

For a fine resolution meter like 17/16, using two rows in the ruler area to print the beat count text will avoid the two digit numbers from running into each other. As a practical matter, print the odd numbers on first row and even number on the second row. Odd tic lines obviously will have to be a bit taller to reach the top row.

**Reply by oddtime** [reply]
> Would a custom sub-division field help solve this dilemma? 

yes that's what I thought

**Reply by theGreatWhiteShark** [reply]
> Would a custom sub-division field help solve this dilemma? The proposal would be to have a field that the grid painter uses. The field would have some comma separated ordered numbers. In the 17/16 example above, one would enter (4,5,4,4) or any other sub-division of choice.

To be honest I do not get this part of the discussion. In terms of labels in the ruler of the pattern editor shouldn't there be the number 1 till 17 printed for a time signature of 17/16? Which one of them is actually displayed will depend on both the zoom factor and the font size as well. Such things are easily automated. We just calculate the maximum width required by one of the numbers and use the overall width of the ruler to determine how many labels can be displayed in total without overlap. When always displaying the first label, which one will follow results automatically.

**Reply by gvergis** [reply]
Although I have neither played, nor likely to play 17/16 in the near future, I thought I would still attempt advancing the discussion along, for the sake of eventually getting to an intuitive 6/8 pattern featured in hydrogen -:). I mostly play 4/4 or 3/4 and occasionally 6/8. So here is an attempt to summarize the 17/16, and generalize the musical meter implementation. If this summary makes sense, I will continue exploring more on the implementation options.

**Reply by gvergis** [reply]
I lost the formatting, so I am attaching a PDF with the information. I hope the attachment stays attached and accessible
[beat group.pdf](https://github.com/hydrogen-music/hydrogen/files/8494041/beat.group.pdf)
.

**Reply by theGreatWhiteShark** [reply]
I like your idea. I'll play a little around and check if and how this can be displayed in the ruler of the pattern editor.

> The choice of count (beat group number vs beat number) option applies to the pattern editor,
director, beat counter widget, and the metronome. Let the user “feel” whatever he chooses.

I can see how this would be useful for e.g. a 17/16 pattern. But I would postpone this feature for later because we are almost done with the next release and I don't want to delay it any longer.

> Beat grouping (how many beats per group) is arbitrary and left to the musician.

Using the "choice of count" feature above, I agree. But in general Hydrogen is not a musical notation software and most users would probably be annoyed if they had to set this on their own. We should provide some reasonable default values.

How about:
- Groups are always quarters (or half notes in the case of 1/4 triplet resolution)
- If the last quarter contains less ticks than half it's size (e.g. 1 for 17/16) it will be dropped and the ticks are merged into the previous group.

**Reply by gvergis** [reply]
Please see the attached 2-page PDF with a possible implementation and GUI mock-up on page 2.
[Hydrogen-Ptn-editor.pdf](https://github.com/hydrogen-music/hydrogen/files/8523873/Hydrogen-Ptn-editor.pdf)

**Reply by gvergis** [reply]
![image](https://user-images.githubusercontent.com/96893274/164355890-895820c2-7bde-49ed-9b81-0ff2f61fa9bd.png)
Slightly modified the mockup of the beat group concept. 

Conditional display of the beat group meter: If "Groups">1 then display beat group meter with the appropriate number of groups, else suppress beat group meter from the pattern editor, altogether.

**Reply by theGreatWhiteShark** [reply]
I see. But these are quite profound changes and we are already in the process of stabilizing the next release. How about we move this to the [issue](https://github.com/hydrogen-music/hydrogen/issues/new) section and make it a new feature request?

**Reply by gvergis** [reply]
Agreed, it is a major undertaking to implement this feature and ultimately provide the "groove feel" by exposing the group parameter in metronome, director, beat counter widget, etc. As for me, I wouldn't use this feature anytime soon. So from my side, there is no issue in waiting. As you suggested, I will make this a new feature request. 

Will the next release support 6/8 pattern properly (the issue that started this long dialog in the first place) ? 😊 

However, I really enjoyed the dialog and opportunity to dig a bit deeper into Hydrogen and rhythm theory. Now that I have dug into this,  I also have some thoughts on "Size", "Res" and "Triplet" feature for the future versions -- I will file this also under the "issues" as a feature request. 

### The grand unification theory 😎

It would be more intuitive and consistent across external applications (admittedly Ardour biased) to use the true musical rhythm "Meter"  like 4/4, 3/4, 6/8, etc. Then the resolution field can be a part of the **sub-division** (AKA: grouplet or Tuplet) concept. I am deliberately using the word sub-division to explicitly associate the process of manipulating the pattern editor's grid. More than merely renaming these fields, I think the overall concept can be made a bit simpler and more intuitive for users as well as developers. At the same time, Hydrogen will become more flexible for any sub-divisions; not just Triplets, but also quintuplet, sextuplet, septuplet, etc.  Once the all-important and fundamental "musical meter" is unambiguously established, the pattern editor's grid resolution can become part of the rhythm **sub-division**  concept. 

Example setting of a sub-division for regular 4/4 meter (no sub-divisions involved) is graphically shown below.
The default pattern editor grid resolution is 1/4 in 4/4 meter. This can be read as "One division at 1/4 interval per note". One division per note could also be interpreted as a note with no sub-divisions. Hence, the resolution defaults to the denominator of the musical meter. Putting it all (Musical Meter and the sub-division widget) together in musical language: In a 4/4 time signature, there are four beats per measure and the quarter note receives one beat. Reading the attached widget parameters left to right, lends to a logical explanation with no ambiguity or contradiction between the musical meter and the sub-division. For a 4/4 meter this seems excessive, however let's look at some other complex meters and see if this overhead is justifiable.

![image](https://user-images.githubusercontent.com/96893274/164944173-2230a66b-182b-491e-91dc-bacb86877f9c.png)

3 1/8th divisions in the time of 2 1/8th notes is shown below -- Alas, the Triplet. We could also have 3 1/4 notes in the time of 2 quarter notes. Rephrasing to make grid sub-division more intuitive: M note Divisions at specified Interval, fit within N notes of the **same interval**.

![image](https://user-images.githubusercontent.com/96893274/164944192-2844b5e9-31e2-4375-9adb-76520a4d2fa6.png)

As we have just seen, this parametric approach starts to pay its dividends quickly by supporting other sub-divisions (hence feature expansion):

Triplet: 3 notes played over the same amount of time as 2 notes of the **same interval** (Graphically shown above)
Quintuplet: 5 notes played over the same amount of time as 4 notes of the **same interval** (Graphically shown below)
Sextuplets: 6 notes played over the same amount of time as 4 notes of the **same interval**
Septuplet: 7 notes played over the same amount of time as 4 notes of the **same interval**

![image](https://user-images.githubusercontent.com/96893274/164936712-47636423-eaa9-46b9-b232-4b904bbc4e12.png)

The grand unification (LOL) happens when we combine sub-divisions with beat grouping. Another words, beat grouping is  highly relevant in the context of higher sub-divisions. They go hand-in-hand, like a pair of matched gloves. As an example: A sextuplet with 6 notes in the time of four become nearly impossible to count at higher tempo. However, counting becomes manageable (or only possible) when the six beats are in two beat groups of three (Counted like Trip-a-let-trip-a-let) or three groups of two. Note that all the note grouping (such as Triplet, Quintuplet, Sextuplet) definitions above, ended with "notes of the **same interval**". This underscores the association of "Interval" (Legacy resolution) to sub-division, such that it should not and cannot be separated. The pattern editor's grid cannot serve two masters! Hence combining the resolution field into this widget such that the pattern's grid sub-division is controlled from the same variable to avoid conflicts and contradictions that arise from two different variables. With this unification theory, a separate resolution field has no place or relevance. Perhaps the sub-division and beat-grouping should also be combined into one widget such that it may be ignored/suppressed (sub-menu or hidden menu) for vast majority of common rhythm users. Just setting the musical meter gives the average user everything they need -- I know I am over simplifying to amplify the prominence of musical meter and its utility as a standalone parameter, for the most basic beat.😊 Most importantly, the length of the measure is indicated by the numerator in beats and the default resolution is indicated by the denominator. Once the measure boundary is framed this way, the sub-divisions concept facilitates the most complex rhythm pattern within the measure through the pattern editor's grid.

To summarize, 1) Musical meter 2) Divisions 3) Interval (resolution field renamed) and 4) notes (may want to call notes something unique like span), are all that is needed to establish any time grid in hydrogen for the purpose of  creating a  beat pattern (the product).  The musical meter is paramount and visible across applications such that they are time synchronized -- its semantics in defining the measure length in Hydrogen's song editor and across other applications should be preserved at all cost. The sub-divisions and beat-grouping are just an aid (template) to establish the beat pattern on a time line (the product) and should have no context outside Hydrogen.

**Reply by theGreatWhiteShark** [reply]
> To summarize, 1) Musical meter 2) Divisions 3) Interval (resolution field renamed) and 4) notes (may want to call notes something unique like span), are all that is needed to establish any time grid in hydrogen for the purpose of creating a beat pattern (the product). The musical meter is paramount and visible across applications such that they are time synchronized -- its semantics in defining the measure length in Hydrogen's song editor and across other applications should be preserved at all cost. The sub-divisions and beat-grouping are just an aid (template) to establish the beat pattern on a time line (the product) and should have no context outside Hydrogen.

I like the idea. In the next release we already have three widgets to control pattern length and grid resolution. But they are numerator and denominator of the former version of the pattern length and the old resolution combo box. That still needs some work.

But before all the tuplet stuff can be properly supported there have to be some (larger) changes in the core of Hydrogen to support float-precision placement of notes. Currently it just works with integer positions. But a major component for this to work was already done for the 1.2 release.

> Will the next release support 6/8 pattern properly (the issue that started this long dialog in the first place) ? blush

Well, it didn't changed too much now the ruler of the pattern editor adjusts depending on the resolution chosen. I hope this is at least a minor improvement :)

![6-8](https://user-images.githubusercontent.com/14164547/166199090-a0b22b83-ec33-419a-9b26-2ca83a91a8c5.png)

**Comment by theGreatWhiteShark** [comment]
The change in the UI of the ruler of the pattern editor makes its insufficiency even more evident. (especially for triplets)

![pic_h2_patternRuler](https://user-images.githubusercontent.com/14164547/160772712-72426b0c-6c4f-40ef-b7d4-42c7107b52ef.png)

The wrong placement of the tick numbers is itself not even a regression. It's just more prominent. I would hate to leave it in this sad state for the 1.2 release.

But I'm not sure yet what to implement. I could write a bit more complex function which will number each and every tick in ruler and takes care of keeping things visible (e.g. rendering only each second or fourth number depending on resolution and zoom factor). But that's probably not what we want. Or just mark quarter triplets instead of quarter notes in case any of the triplet choices is picked?

**Reply by gvergis** [reply]
I agree with the complications that the "compound meters" such as 6/8 introduces on the  existing pattern editor. Trying to implement everything discussed on the thread will require some deeper thought and significant simplification. I have started to think about this from two perspectives: 1) Musical theory and musicality (count/feel)** within hydrogen 2) Transport synchronization and consistency with other music applications. ** Measure and beat count must be correctly displayed in the "director" window in hydrogen, consistent with the settings in pattern/song editor  (this is one of my most favorite features in Hydrogen)

As a matter of fact, I have to brush up on some rhythm theory, myself. I play music intuitively, without over analyzing the theory behind! So, this is a good learning opportunity for me as well; So, please be patient with me as I bounce ideas back and forth and turning intuition into theory and algorithm. Stay tuned...
